src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -0,0 +1,728 @@
+    analyzer = from_.analyzer
+    column_states = ColumnStateDict()
+    column_index = 0
+    # populate column status against subquery
+    quoted_col_names = []
+    for c in cols:
+        if isinstance(c, UnresolvedAlias) and isinstance(c.child, Star):
+            column_states.update(from_.column_states)
+            continue
+        c_name = parse_column_name(c, analyzer)
+        quoted_c_name = analyzer_utils.quote_name(c_name)
+        quoted_col_names.append(quoted_c_name)
+        from_c_state = from_.column_states.get(quoted_c_name)
+        if from_c_state and from_c_state.change_state != ColumnChangeState.DROPPED:
+            if c_name != from_.analyzer.analyze(c):
+                column_states[quoted_c_name] = ColumnState(
+                    quoted_c_name,
+                    column_index,
+                    ColumnChangeState.CHANGED_EXP,
+                    c,
+                    state_dict=column_states,
+                )
+            else:
+                column_states[quoted_c_name] = ColumnState(
+                    quoted_c_name,
+                    column_index,
+                    ColumnChangeState.UNCHANGED_EXP,
+                    c,
+                    state_dict=column_states,
+                )
+        else:
+            column_states[quoted_c_name] = ColumnState(
+                quoted_c_name,
+                column_index,
+                ColumnChangeState.NEW,
+                c,
+                state_dict=column_states,
+            )
+        column_index += 1
+    # end of populate column status against subquery
+
+    # populate column dependency
+    for c, quoted_c_name in zip(cols, quoted_col_names):
+        dependent_column_names = get_dependent_columns(c)
+        column_states[quoted_c_name].dependent_columns = dependent_column_names
+        if dependent_column_names == COLUMN_DEPENDENCY_DOLLAR:
+            column_states[quoted_c_name].depend_on_same_level = False
+        elif dependent_column_names == COLUMN_DEPENDENCY_ALL:
+            column_states[quoted_c_name].depend_on_same_level = True
+            column_states.columns_referencing_all_columns.add(quoted_c_name)
+        else:
+            for dependent_column in dependent_column_names:
+                if dependent_column not in from_.column_states.active_columns:
+                    column_states[quoted_c_name].depend_on_same_level = True
+                    if dependent_column in column_states:
+                        column_states[
+                            dependent_column
+                        ].add_referenced_by_same_level_column(dependent_column)
+                    else:  # A referenced column can't be found.
+                        raise DeriveColumnDependencyError()
+    # end of populate column dependency
+
+    for dc in from_.column_states.active_columns - column_states.active_columns:
+        # for dropped columns, we only care name.
+        column_states[dc] = ColumnState(
```

Comment:
We probably should have factory methods for generating `ColumnState` so we don't need to do this many `None`.